File: android/src/main/java/com/lightningkite/kwift/observables/WriteAddedObservableProperty.shared.kt
package com.lightningkite.kwift.observables

import com.lightningkite.kwift.Box
import com.lightningkite.kwift.escaping
import io.reactivex.Observable

class WriteAddedObservableProperty<A>(
    val basedOn: ObservableProperty<A>,
    val onWrite: @escaping() (A) -> Unit
) : MutableObservableProperty<A>() {
    override var value: A
        get() = basedOn.value
        set(value) {
            onWrite(value)
        }
    override val onChange: Observable<Box<A>> get() = basedOn.onChange
    override fun update() {
        //Do nothing
    }
}

fun <T> ObservableProperty<T>.withWrite(
    onWrite: @escaping() (T) -> Unit
): MutableObservableProperty<T> {
    return WriteAddedObservableProperty<T>(this, onWrite)
}

File: ios/Kwift/com/lightningkite/kwift/observables/WriteAddedObservableProperty.shared.swift
//Package: com.lightningkite.kwift.observables
//Converted using Kwift2

import Foundation



public class WriteAddedObservableProperty<A>: MutableObservableProperty<A> {
    
    public var basedOn: ObservableProperty<A>
    public var onWrite:  (A) -> Void
    
    
    override public func update() -> Void {
        onWrite(basedOn.value)
    }
    override public var value: A {
        get {
            return basedOn.value
        }
        set(value) {
            onWrite(value)
        }
    }
    override public var onChange: Event<A> {
        get {
            return basedOn.onChange
        }
    }
    
    public init(basedOn: ObservableProperty<A>, onWrite: @escaping (A) -> Void) {
        self.basedOn = basedOn
        self.onWrite = onWrite
        super.init()
    }
    convenience public init(_ basedOn: ObservableProperty<A>, _ onWrite: @escaping (A) -> Void) {
        self.init(basedOn: basedOn, onWrite: onWrite)
    }
}
 
 

extension ObservableProperty {
    public func withWrite(onWrite: @escaping (T) -> Void) -> MutableObservableProperty<T> {
        return WriteAddedObservableProperty<T>(self, onWrite)
    }
}
 
